core/save-git.c

```diff
@@ -1328,43 +1331,66 @@ int do_git_save(git_repository *repo, const char *branch, const char *remote, bo
 	 * Check if we can do the cached writes - we need to
 	 * have the original git commit we loaded in the repo
 	 */
-	cached_ok = try_to_find_parent(saved_git_id, repo);
+	cached_ok = try_to_find_parent(saved_git_id, info->repo);
 
 	/* Start with an empty tree: no subdirectories, no files */
 	tree.name[0] = 0;
 	tree.subdirs = NULL;
-	if (git_treebuilder_new(&tree.files, repo, NULL))
+	if (git_treebuilder_new(&tree.files, info->repo, NULL))
 		return report_error("git treebuilder failed");
 
 	if (!create_empty)
 		/* Populate our tree data structure */
-		if (create_git_tree(repo, &tree, select_only, cached_ok))
+		if (create_git_tree(info->repo, &tree, select_only, cached_ok))
 			return -1;
 
 	if (verbose)
 		SSRF_INFO("git storage, write git tree\n");
 
-	if (write_git_tree(repo, &tree, &id))
+	if (write_git_tree(info->repo, &tree, &id))
 		return report_error("git tree write failed");
 
 	/* And save the tree! */
-	if (create_new_commit(repo, remote, branch, &id, create_empty))
+	if (create_new_commit(info, &id, create_empty))
 		return report_error("creating commit failed");
 
 	/* now sync the tree with the remote server */
-	if (remote && !git_local_only)
-		return sync_with_remote(repo, remote, branch, url_to_remote_transport(remote));
+	if (info->url && !git_local_only)
+		return sync_with_remote(info);
 	return 0;
 }
 
-int git_save_dives(struct git_repository *repo, const char *branch, const char *remote, bool select_only)
+int git_save_dives(struct git_info *info, bool select_only)
 {
 	int ret;
 
-	if (repo == dummy_git_repository)
-		return report_error("Unable to open git repository '%s'", branch);
-	ret = do_git_save(repo, branch, remote, select_only, false);
-	git_repository_free(repo);
-	free((void *)branch);
+	/*
+	 * FIXME!! This open_git_repository() will
+	 * sync with the cloud. That is NOT what
+	 * we actually want to do here. We want
+	 * to just open the local repo, and then
+	 * sync with the clound after having saved.
+	 *
+	 * NOTE! The "we already have a repo" case
+	 * case is fairly easy (we can just treat
+	 * it as a local repository and open it
+	 * as-is). But the "need to create it by
+	 * cloning" case is what we currently also
+	 * do in 'open_git_repository()', and that
+	 * all needs to be thought about a lot.
+	 *
+	 * Do we want to have the rule that you
```

Comment:
This is the meat of the question. 

In many ways, it really would make sense to only save to a git repo that we opened earlier - if for no other reason than the fact that we need to have the parent commit. So the whole "create initial empty repo" is possibly only an issue at load time, and save-time could assume that a simple "git_repository_open(&info->repo, info->localdir)" should be sufficient.

Even if that's the case, we'd still want to try to push out those changes after writing, so it's not *just* about changing a "open_git_repository()" to "git_repository_open()".

Reply:
I think the best approach to this is to think about the different scenarios that we need to deal with.
And that's surprisingly complicated for something that seems so simple...

(1) this system has previously connected to the cloud and has a local repo that's in sync with the remote
(2) this system has previously connected to the cloud and has a local repo that is ahead of the remote
(3) this system has previously connected to the cloud and has a local repo that's behind the remote
(4) this system has previously connected to the cloud and has a local repo that has diverged from the remote

So in cases 1-4 the local and remote repos have a common ancestor, and neither (1), the local repo (2), the remote repo (3) or both (4) have commits since then

(5) this system has previously been in no-cloud mode and has a local repo for that, but no common ancestor with the remote
(6) this system is trying to connect to the cloud for the first time, the remote repo is brand new (i.e., doesn't have the branch)
(7) this system is trying to connect to the cloud for the first time, the remote repo exists and has commits in it
(8) various weird error cases where we have issues with the local or remote repo

Additionally, there are a few different situations that we can be in locally
(A) first time use with these credentials - unknown if they are correct
(B) this system has known good credentials and wants to get new content from the cloud (i.e. the user started Subsurface with auto-sync or started Subsurface and did a manual sync)
(B) this system has changes that need to be stored to the local repo
(C) this system has changes in the local report that need to be pushed to the remote repo

And then you can build a little matrix with the flow that is needed. And in each case you need to account for the possibility that you can't connect to the cloud (what's the end state?) or that there is an issue merging or that there is some other error...

Are we having fun, yet?

Fundamentally this complexity is the reason for the utter nightmare of code that we have here. A few of these cases got added later - and I know that we are not doing the right thing for some of them today.
But I think the more we try to create a clear diagram of the different states and transitions and possible outcomes, the easier it will be to make sure that we get it right.

Oh, and I'm not even 100% sure that the list above is complete :)